lib/qbo_api/api_methods.rb

```diff
@@ -67,6 +67,97 @@ def deactivate(entity, id:)
       request(:post, entity: entity, path: entity_path(entity), payload: payload)
     end
 
+    # https://developer.intuit.com/docs/api/accounting/invoice
+    # Get an invoice as PDF
+    # Operation:
+    #
+    #   GET /v3/company/<realmID>/invoice/<invoiceId>/pdf
+    #   Content type: application/pdf
+    #
+    # This resource returns the specified object in the response body as
+    # an Adobe Portable Document Format (PDF) file. The resulting PDF file is
+    # formatted according to custom form styles in the company settings.
+    #    %PDF-1.4
+    #    ...
+    #    %%EOF
+    def get_invoice_pdf(invoice_id)
+      headers = { 'Content-Type' => 'application/pdf', 'Accept' => 'application/pdf' }
+      connection = build_connection(@endpoint_url, headers: headers) do |conn|
+        add_exception_middleware(conn)
+        add_authorization_middleware(conn)
+        add_connection_adapter(conn)
+      end
+      path = "#{entity_path(:invoice)}/#{invoice_id}/pdf"
+      raw_response = raw_request(:get, conn: connection, path: path)
+      raw_response.body
+    end
+
+    # https://developer.intuit.com/docs/api/accounting/invoice
+    # Operation:
+    #
+    # Using email address supplied in Invoice.BillEmail.EmailAddress:
+    #
+    # POST /v3/company/<realmID>/invoice/<invoiceId>/send
+    #   Content type: application/octet-stream
+    #
+    # Specifying an explicit email address:
+    #
+    # POST /v3/company/<realmID>/invoice/<invoiceId>/send?sendTo=<emailAddr>
+    #   Content type: application/octet-stream
+    #
+    # Once email is sent, the following actions happen:
+    #
+    #    The Invoice.EmailStatus parameter is set to EmailSent.
+    #    The Invoice.DeliveryInfo element is populated with sending information<./li>
+    #    The Invoice.BillEmail.Address parameter is updated to the address specified with the value of the sendTo query parameter, if specified.
+    # For example
+    #       "EmailStatus": "EmailSent",
+    #       "DeliveryInfo":
+    #            {
+    #                "DeliveryType": "Email",
+    #                "DeliveryTime": "2014-12-17T11:50:52-08:00"
+    #            },
+    #        "BillEmail": {
+    #          "Address": "test@intuit.com"
+    #        },
+    def deliver_invoice(invoice_id, send_to: nil, params: nil)
+      headers = {
+        'Accept' => 'application/json;charset=UTF-8',
+        'Content-Type' => 'application/octet-stream'
+      }
+      connection = build_connection(@endpoint_url, headers: headers) do |conn|
+        add_authorization_middleware(conn)
+        add_exception_middleware(conn)
+        add_connection_adapter(conn)
+      end
+      params ||= {}
+      params[:sendTo] = send_to if send_to
+      path = "#{entity_path(:invoice)}/#{invoice_id}/send"
+      raw_response = raw_request(:get, conn: connection, path: path, params: params)
```

Comment:
bug: is post

I need to see also if this method lets me easily create a proxy tax agency, proxy tax code, and proxy tax rate

my notes:


per https://developer.intuit.com/docs/api/accounting/invoice

> If `ShipAddr`, `BillAddr`, or both are not provided, the appropriate customer address from the referenced `Customer` object is used to fill those values.

> `ShipAddr` Identifies the address where the goods must be shipped. If not specified, and a default `Customer:ShippingAddr` is specified in QuickBooks for this customer, the default ship-to address will be used by QuickBooks.

per https://developer.intuit.com/hub/blog/2017/12/11/using-quickbooks-online-api-automated-sales-tax

> a company is automated tax service enabled if ` Preferences.TaxRefs.PartnerTaxEnabled` is true and sales tax is set up. ( Sandbox support is expected around January 15, 2018 )
>  Sales tax is computed automatically on the sales transaction based upon the shipping address `ShipAddr` and the location of the company `CompanyInfo.LegalAddr` rather assigning sales tax manually.
> Tax is determined based on the source and destination address. 
>  The source address is the company’s legal address `CompanyInfo.LegalAddr` as available in the company settings.  The ability to customize the source address based on location of a given transaction is not supported.
> The destination address is the shipping address `ShipAddr` provided on the sales transaction. If a shipping address `ShipAddr` is not provided, the company address ``CompanyInfo.LegalAddr` is considered as the destination address. 
> For more details see https://community.intuit.com/articles/1567657-automated-sales-tax

|                                             | QBO UI |  QBO API |
|--------------------------------|-----------|------------|
| Customer default tax code | No longer available or visible. | Customer.DefaultTaxCodeRef populated with company’s default tax code as defined by AST.
| Company default tax code | Set via AST based on the tax agency. | TaxCode object created by AST.
| Company default tax agency | AST sets the default tax agency. Subsequent ones can be added manually. | TaxAgency object created by AST.
| Create operation available for TaxCode objects | No. | Can create and assign on transactions but will be overridden by AST. Create operation exists for backward compatibility with existing apps and can be queried.

> creating TaxCode, TaxRate, and TaxAgency objects is still possible for backward compatibility to support existing apps. They can be accessed by the API but are not visible in the UI.
> The app can include a user or API created tax code in the transaction request payload but AST overrides the code with its recommendation.
> An app can still attach a TaxCode object to a transaction, but the automated tax service replaces an incorrect TaxCode object with one that is correct for the transaction.

> At the moment, the app cannot present the AST-recommended sales tax to the user until the transaction is first saved to the QuickBooks Online company. 
>  The following process is recommended for apps implementing V3 of the API:
> 1) Create a TaxCode object via API using the Taxservice resource that serves as a proxy, named Sales Tax Code, with an arbitrary percentage like 5%. This is not visible on the QBO Sales Tax screen, but can still be queried via API.
> 2) Associate this proxy TaxCode object to all transactions that are posted to QuickBooks Online.
> 3) QuickBooks Online services swaps out the proxy TaxCode for the AST-recommended TaxCode object.
> 4) Thereafter, when the transaction is read back, the right sales tax is returned.


> Considerations for existing integrations

| Use case | Tax objects used in the transaction request payload | Pre-AST behavior | AST behavior |
| -- | -- | -- | -- |
| Specifying sales tax | TxnTaxDetail.TxnTaxCodeRef | Tax amount calculated based on associated tax rates in the tax code. | Ignore TaxCode passed in and calculate tax based on AST-assigned tax code.TxnTaxDetail.TxnTaxCodeRef in response payload is changed to AST-recommended tax code.
| Overriding sales tax | TxnTaxDetail.TxnTaxCodeRefTxnTaxDetail.TotalTax | Prorates supplied tax amount across associated tax rates in the tax code. | TxnTaxDetail.TotalTaxamount is honored. Supplied tax amount is prorated across associated tax rates in the AST-assigned tax code. TxnTaxDetail.TxnTaxCodeRefin response payload is changed to AST-assigned tax code.

ref
- https://developer.intuit.com/docs/00_quickbooks_online/2_build/60_tutorials/0061_manage_sales_tax_for_us_locales
- https://developer.intuit.com/docs/api/accounting/taxcode https://developer.intuit.com/docs/api/accounting/taxservice  https://developer.intuit.com/docs/api/accounting/taxagency